File: src/game/engine.ts
/**
 * Pure rules engine for the color-sort puzzle. Every function here is side-effect free
 * and immutable: `pour` returns a brand new `GameState` rather than mutating its input,
 * which keeps undo history and React rendering simple.
 */
import type { Bottle, Color, GameState, Move } from './types';

/** The top (pourable) color of a bottle, or `null` if the bottle is empty. */
export function topColor(bottle: Bottle): Color | null {
  return bottle.length > 0 ? bottle[bottle.length - 1]! : null;
}

/** How many contiguous segments of the same color sit at the top of the bottle. */
export function topRunLength(bottle: Bottle): number {
  if (bottle.length === 0) return 0;
  const color = bottle[bottle.length - 1]!;
  let run = 1;
  for (let i = bottle.length - 2; i >= 0; i--) {
    if (bottle[i] === color) run++;
    else break;
  }
  return run;
}

/** Empty slots remaining in a bottle. */
export function freeSpace(bottle: Bottle, capacity: number): number {
  return capacity - bottle.length;
}

/** True when a bottle is empty, or full and a single solid color. */
export function isComplete(bottle: Bottle, capacity: number): boolean {
  if (bottle.length === 0) return true;
  if (bottle.length !== capacity) return false;
  const first = bottle[0];
  return bottle.every((c) => c === first);
}

/**
 * Whether a pour from `from` to `to` is legal:
 * different bottles, source non-empty, target has room, and the colors match
 * (or the target is empty).
 */
export function canPour(state: GameState, from: number, to: number): boolean {
  if (from === to) return false;
  const src = state.bottles[from];
  const dst = state.bottles[to];
  if (!src || !dst) return false;
  if (src.length === 0) return false;
  if (freeSpace(dst, state.capacity) <= 0) return false;
  const dstTop = topColor(dst);
  return dstTop === null || dstTop === topColor(src);
}

/**
 * How many segments would actually move on a `from -> to` pour: the smaller of the
 * source's top run and the target's free space. Returns 0 if the pour is illegal.
 */
export function pourAmount(state: GameState, from: number, to: number): number {
  if (!canPour(state, from, to)) return 0;
  const src = state.bottles[from]!;
  const dst = state.bottles[to]!;
  return Math.min(topRunLength(src), freeSpace(dst, state.capacity));
}

/**
 * Apply a pour, returning the resulting state and the executed move.
 * Throws if the pour is illegal — callers should gate with `canPour` first.
 *
 * `maxCount` caps how many segments move (default: the full top run). The hidden-colors
 * mechanic uses it to limit a pour to the player-visible run; the engine itself stays
 * unaware of concealment.
 */
export function pour(
  state: GameState,
  from: number,
  to: number,
  maxCount = Infinity,
): { state: GameState; move: Move } {
  const count = Math.min(pourAmount(state, from, to), maxCount);
  if (count <= 0) {
    throw new Error(`Illegal pour from ${from} to ${to}`);
  }
  const src = state.bottles[from]!;
  const dst = state.bottles[to]!;
  const color = topColor(src)!;

  const newSrc = src.slice(0, src.length - count);
  const newDst = [...dst, ...Array<Color>(count).fill(color)];

  const bottles = state.bottles.map((bottle, i) => {
    if (i === from) return newSrc;
    if (i === to) return newDst;
    return bottle;
  });

  return { state: { ...state, bottles }, move: { from, to, count, color } };
}

/** A level is won when every bottle is complete (empty or a single full color). */
export function isWon(state: GameState): boolean {
  return state.bottles.every((b) => isComplete(b, state.capacity));
}

/** Enumerate every legal pour available from a state. */
export function legalMoves(state: GameState): Array<{ from: number; to: number }> {
  const moves: Array<{ from: number; to: number }> = [];
  const n = state.bottles.length;
  for (let from = 0; from < n; from++) {
    for (let to = 0; to < n; to++) {
      if (canPour(state, from, to)) moves.push({ from, to });
    }
  }
  return moves;
}

/** Deadlock: not won, and no legal move remains (the "no moves left" alert). */
export function isDeadlocked(state: GameState): boolean {
  return !isWon(state) && legalMoves(state).length === 0;
}

File: core/src/engine.rs
//! Pure rules engine — the line-for-line port of `src/game/engine.ts`. Same immutable
//! semantics (a pour returns a new state); enumeration order (from asc × to asc) is preserved
//! exactly because the DFS solver's byte-reproducible solutions depend on it.

use crate::state::{State, Tube};
use crate::types::Move;

/// Contiguous same-color run at the top of a tube.
pub fn top_run_length(t: &Tube) -> usize {
    let n = t.len();
    if n == 0 {
        return 0;
    }
    let color = t.cell(n - 1);
    let mut run = 1;
    for i in (0..n - 1).rev() {
        if t.cell(i) == color {
            run += 1;
        } else {
            break;
        }
    }
    run
}

#[inline]
pub fn free_space(t: &Tube, capacity: u8) -> usize {
    capacity as usize - t.len()
}

/// Empty, or full and a single solid color.
pub fn is_complete(t: &Tube, capacity: u8) -> bool {
    if t.is_empty() {
        return true;
    }
    if t.len() != capacity as usize {
        return false;
    }
    t.is_uniform()
}

pub fn can_pour(state: &State, from: usize, to: usize) -> bool {
    if from == to {
        return false;
    }
    let (Some(src), Some(dst)) = (state.tubes.get(from), state.tubes.get(to)) else {
        return false;
    };
    if src.is_empty() {
        return false;
    }
    if free_space(dst, state.capacity) == 0 {
        return false;
    }
    match dst.top() {
        None => true,
        top => top == src.top(),
    }
}

/// Segments a `from -> to` pour would move; 0 if illegal.
pub fn pour_amount(state: &State, from: usize, to: usize) -> usize {
    if !can_pour(state, from, to) {
        return 0;
    }
    top_run_length(&state.tubes[from]).min(free_space(&state.tubes[to], state.capacity))
}

/// Apply a pour capped at `max_count` segments, returning the new state and executed move.
/// Panics if illegal — callers gate with `can_pour`, exactly like the JS contract (`pour`
/// throws). `usize::MAX` = uncapped (the JS `Infinity` default).
pub fn pour(state: &State, from: usize, to: usize, max_count: usize) -> (State, Move) {
    let count = pour_amount(state, from, to).min(max_count);
    assert!(count > 0, "illegal pour from {from} to {to}");
    let color = state.tubes[from].top().unwrap();

    let mut next = state.clone();
    next.tubes[from].pop_n(count);
    next.tubes[to].push_n(color, count);

    (next, Move { from: from as u8, to: to as u8, count: count as u8, color })
}

/// Won when every tube is complete (empty or a single full color).
pub fn is_won(state: &State) -> bool {
    state.tubes.iter().all(|t| is_complete(t, state.capacity))
}

/// Every legal pour, in the canonical (from asc, to asc) order the solver's determinism
/// depends on.
pub fn legal_moves(state: &State) -> Vec<(u8, u8)> {
    let n = state.tubes.len();
    let mut moves = Vec::new();
    for from in 0..n {
        for to in 0..n {
            if can_pour(state, from, to) {
                moves.push((from as u8, to as u8));
            }
        }
    }
    moves
}

/// Not won and no legal move remains.
pub fn is_deadlocked(state: &State) -> bool {
    !is_won(state) && legal_moves(state).is_empty()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn state(tubes: Vec<&[u8]>, capacity: u8) -> State {
        State { tubes: tubes.into_iter().map(Tube::from_cells).collect(), capacity }
    }

    #[test]
    fn pour_moves_the_top_run_up_to_free_space() {
        let s = state(vec![&[0, 1, 1], &[1], &[]], 4);
        assert_eq!(pour_amount(&s, 0, 1), 2);
        let (next, mv) = pour(&s, 0, 1, usize::MAX);
        assert_eq!(mv, Move { from: 0, to: 1, count: 2, color: 1 });
        assert_eq!(next.tubes[0].cells(), &[0]);
        assert_eq!(next.tubes[1].cells(), &[1, 1, 1]);
    }

    #[test]
    fn capped_pour_respects_max_count() {
        let s = state(vec![&[1, 1, 1], &[]], 4);
        let (next, mv) = pour(&s, 0, 1, 1);
        assert_eq!(mv.count, 1);
        assert_eq!(next.tubes[0].cells(), &[1, 1]);
    }

    #[test]
    fn win_and_deadlock() {
        assert!(is_won(&state(vec![&[2, 2, 2, 2], &[]], 4)));
        // Full mixed tubes, no free space: deadlocked.
        let s = state(vec![&[0, 1, 0, 1], &[1, 0, 1, 0]], 4);
        assert!(is_deadlocked(&s));
    }
}
